File: qeli-android/app/src/main/kotlin/com/qeli/TrafficShaper.kt
package com.qeli

import kotlin.math.ln
import kotlin.math.max
import kotlin.random.Random

/**
 * Idle cover-traffic scheduler — the Kotlin mirror of the Rust `protocol::shaper`
 * (DPI-AUDIT 6.1/6.2). When enabled, an idle tunnel emits cover packets at gaps
 * sampled from an exponential (Poisson-process) distribution rather than a fixed
 * heartbeat, with a browsing-ish size distribution, capped by a byte budget.
 * Cover packets are empty-payload encrypted records the peer drops, so this is
 * not a wire-format change. Sampling is timing/size only (not secret).
 */
class TrafficShaper(
    enabledIn: Boolean,
    private val gapMeanMs: Long,
    private val gapMinMs: Long,
    gapMaxMs: Long,
    private val budgetBytesPerSec: Int,
    private val minSize: Int,
    maxSize: Int,
    stealthIn: Boolean = false,
    stealthRateMbps: Int = 2,
) {
    val enabled: Boolean = enabledIn && budgetBytesPerSec > 0

    /** Stealth: rate-cap the data plane + cover under load. Implies [enabled].
     * TCP-only (the caller gates UDP off, mirroring the Rust core). */
    val stealth: Boolean = enabled && stealthIn

    private val gapMax: Long = max(gapMinMs, gapMaxMs)
    private val sizeMax: Int = max(minSize, maxSize)
    private val stealthRateBps: Double = max(1, stealthRateMbps) * 1_000_000.0
    private var tokens: Double = budgetBytesPerSec.toDouble()
    private var lastRefillNanos: Long = System.nanoTime()
    // Separate token bucket (bits) for the stealth data-plane rate cap.
    private var rateTokens: Double = 0.0
    private var rateLastNanos: Long = System.nanoTime()

    /** Stealth data-plane pacing: account [bytes] against the rate cap and return how
     * long (ms) to sleep before the next send (0 if under budget or stealth is off).
     * Carries a deficit so bursts average to the cap. */
    fun stealthPaceMs(bytes: Int): Long {
        if (!stealth) return 0
        val now = System.nanoTime()
        val elapsed = (now - rateLastNanos) / 1_000_000_000.0
        rateLastNanos = now
        rateTokens = minOf(rateTokens + elapsed * stealthRateBps, stealthRateBps)
        rateTokens -= bytes * 8.0
        return if (rateTokens >= 0) 0 else minOf(1000.0, -rateTokens / stealthRateBps * 1000.0).toLong()
    }

    /** Next inter-cover gap (ms): exponential (inverse-CDF), clamped to [min,max]. */
    fun nextGapMs(): Long {
        val u = Random.nextDouble()
        val sampled = -max(1L, gapMeanMs).toDouble() * ln(max(1e-12, 1.0 - u))
        return sampled.toLong().coerceIn(gapMinMs, gapMax)
    }

    /** Sample a cover packet size in [minSize, maxSize]. */
    fun nextSize(): Int = if (minSize >= sizeMax) minSize else Random.nextInt(minSize, sizeMax + 1)

    /** Token-bucket check+spend; true (and deducts) if the budget allows [bytes]. */
    fun trySpend(bytes: Int): Boolean {
        if (budgetBytesPerSec <= 0) return false
        val now = System.nanoTime()
        val elapsed = (now - lastRefillNanos) / 1_000_000_000.0
        lastRefillNanos = now
        tokens = minOf(tokens + elapsed * budgetBytesPerSec, budgetBytesPerSec.toDouble())
        return if (tokens >= bytes) { tokens -= bytes; true } else false
    }
}

File: qeli-shared/QeliShared/Protocol/TrafficShaper.cs
using System.Diagnostics;

namespace Qeli.Shared.Protocol;

/// <summary>
/// Idle cover-traffic scheduler — the C# mirror of the Rust <c>protocol::shaper</c>
/// (DPI-AUDIT 6.1/6.2). When enabled, an idle tunnel emits cover packets at gaps
/// sampled from an exponential (Poisson-process) distribution rather than a fixed
/// heartbeat, with a browsing-ish size distribution, capped by a byte budget.
/// Cover packets are empty-payload encrypted records the peer drops, so this is
/// not a wire-format change. Sampling is timing/size only (not secret), so the
/// system <see cref="Random"/> is fine here.
/// </summary>
public sealed class TrafficShaper
{
    public bool Enabled { get; }

    private readonly double _gapMeanMs;
    private readonly long _gapMinMs;
    private readonly long _gapMaxMs;
    private readonly int _budgetBytesPerSec;
    private readonly int _minSize;
    private readonly int _maxSize;
    private readonly Random _rng = new();
    private double _tokens;
    private long _lastRefillTicks;

    public TrafficShaper(bool enabled, long gapMeanMs, long gapMinMs, long gapMaxMs,
                         int budgetBytesPerSec, int minSize, int maxSize)
    {
        Enabled = enabled && budgetBytesPerSec > 0;
        _gapMeanMs = Math.Max(1, gapMeanMs);
        _gapMinMs = gapMinMs;
        _gapMaxMs = Math.Max(gapMinMs, gapMaxMs);
        _budgetBytesPerSec = budgetBytesPerSec;
        _minSize = minSize;
        _maxSize = Math.Max(minSize, maxSize);
        _tokens = budgetBytesPerSec; // start with ~1s of budget
        _lastRefillTicks = Stopwatch.GetTimestamp();
    }

    /// <summary>Next inter-cover gap in ms — exponential (inverse-CDF of Exp(1/mean)),
    /// clamped to [min, max]. The exponential tail is what makes it non-periodic.</summary>
    public int NextGapMs()
    {
        double u = _rng.NextDouble();
        double sampled = -_gapMeanMs * Math.Log(Math.Max(1e-12, 1.0 - u));
        return (int)Math.Clamp(sampled, _gapMinMs, _gapMaxMs);
    }

    /// <summary>Sample a cover packet size in [minSize, maxSize].</summary>
    public int NextSize() => _minSize >= _maxSize ? _minSize : _rng.Next(_minSize, _maxSize + 1);

    /// <summary>Token-bucket check+spend; true (and deducts) if the budget allows
    /// <paramref name="bytes"/> of cover, false if over budget (skip this cover).</summary>
    public bool TrySpend(int bytes)
    {
        if (_budgetBytesPerSec <= 0) return false;
        long now = Stopwatch.GetTimestamp();
        double elapsed = (now - _lastRefillTicks) / (double)Stopwatch.Frequency;
        _lastRefillTicks = now;
        // Cap at ~1s of budget so an idle period can't bank a burst.
        _tokens = Math.Min(_tokens + elapsed * _budgetBytesPerSec, _budgetBytesPerSec);
        if (_tokens >= bytes) { _tokens -= bytes; return true; }
        return false;
    }
}
